### PR TITLE
Adding in-memory nexus file and adding components to instrument group

### PR DESCRIPTION
**Pull request by rerpha**
### Issue

Refs #108 

### Description of work

Adding a separate model for creating an in-memory nexus file. This currently doesn't really do anything, and doesn't add the sample, geometry, pixel info or any transformations to the file, but it does add components when created to the `/entry/instrument` group. 

I have added some tests to cover this functionality, as well as some tests to cover the new model. 

Moved `ComponentType` into separate file for testing.

### Acceptance Criteria 

Tests should cover what i have changed, but it would be good to make sure nothing has broken. 

### Nominate for Group Code Review

- [ ] Nominate for code review

### CODE REVIEW

File: nexus_constructor/qml_models/instrument_model.py [hunk 1]
```diff
@@ -210,6 +228,9 @@ def add_component(
                 if transform_model is None
                 else transform_model.transforms,
             )
+            create_group(
```
Comment (matthew-d-jones) [inline nexus_constructor/qml_models/instrument_model.py]: Should the `Component` be responsible for this?

**Comment by rerpha** [comment]
I believe this is now ready to go in. It's the first part of adding the nexus model so doesn't really do anything but it's a start.

File: nexus_constructor/qml_models/instrument_model.py [hunk 2]
```diff
@@ -69,23 +67,48 @@ class InstrumentModel(QAbstractListModel):
...
+        sample = create_component(
+            component_type=ComponentType.SAMPLE,
+            name="Sample",
+            geometry=OFFCube,
+            parent_group=group,
+        )
+
+        self.create_instrument_group(group)
+
+        self.beginInsertRows(QModelIndex(), self.rowCount(), self.rowCount())
```
[the remark this answers is not in the document]
Comment (DominicOram) [reply nexus_constructor/qml_models/instrument_model.py]: Also, this whole block of code is incredibly similar as in create_component, can we refactor please?

[the remark this answers is not in the document]
Comment (DominicOram) [reply nexus_constructor/qml_models/instrument_model.py]: Can you add a bit more documentation about why this method exists and we can't do it all in the constructor.

**Comment by rerpha** [comment]
> I've added some specific things in the code but also are we sure we have to have this weird initialisation? Do we have to create the nexus model in the QML?

we could create the nexus model in the instrument model, however i'm not keen to do that as currently the instrumentModel contains components and might not be suited for everything else in the nexus file. There is stuff outside the nexus file that the instrument model shouldn't be responsible for

On the other side we could rename InstrumentModel to NexusModel if it is going to be responsible for much more than just components. If that's the case we need to see if there's a better QML data type for it as it will be more like a tree structure.

File: nexus_constructor/qml_models/instrument_model.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rerpha) [reply nexus_constructor/qml_models/instrument_model.py]: I'm not sure how to create a context manager within the scope of a class which references the slots like beginInsertRows, but that would be nice. If that is done we wouldn't need to refactor it

**Comment by rerpha** [comment]
Have created #240 for the context manager stuff that dom mentioned. I think this will apply to everything else that uses QAbstractLists such as transformModel and pixelMappingModel

**Pull request merged**